include/drivers/hwtrig.h

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2021 Shlomi Vaknin
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public HWTRIG Driver APIs
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_HWTRIG_H_
+#define ZEPHYR_INCLUDE_DRIVERS_HWTRIG_H_
+
+/**
+ * @brief HWTRIG Interface
+ * @defgroup hwtrig_interface HWTRIG Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <errno.h>
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <sys/math_extras.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Provides a type to hold HWTRIG configuration flags.
+ */
+typedef void* hwtrig_flags_t;
+
+/**
+ * @typedef hwtrig_enable_t
+ * @brief Enable the hardware trigger
+ * See @a hwtrig_enable() for argument description
+ */
+typedef int (*hwtrig_enable_t)(const struct device *dev,
+                    hwtrig_flags_t flags);
+
+/** @brief HWTRIG driver API definition. */
+__subsystem struct hwtrig_driver_api {
+	hwtrig_enable_t enable;
+};
+
+/**
+ * @brief Enable HWTRIG trigger.
+ *
+ * This function enables the hardware trigger configuration. What the
+ * configuration is should be driver specific.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param flags trigger configuration.
+ *
+ * @retval 0 If successful.
+ * @retval -EINVAL if invalid function parameters were given
+ */
+__syscall int hwtrig_enable(const struct device *dev, hwtrig_flags_t flags);
```

Comment:
I think the goal of an API should be to abstract the hardware. Ideally, I would write an application for STM32 that uses this API and I could use exactly the same code also for other MCUs. So I don't think we can pass a void pointer here where we don't know what it is used for in the driver and expect the user to look at the driver code.

The frequency of the trigger is probably something that has to be set in any case, so it might be explicitly part of the API.

Reply:
I agree. This api is not abstracting the hardware very well. The reason is that I couldn't come up with a proper solution to the fact that there might be some triggers that are not timer based, and not have `sampling_frequency` property. If we are dealing with timer triggers, than I would probably use `set_sampling_frequency` function and remove the `flags` parameter in the `enable` function (and maybe add `disable` function as well).